pkg/image/prune/imagepruner.go

```diff
@@ -0,0 +1,744 @@
+package prune
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+
+	kapi "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	kclient "github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/fields"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/golang/glog"
+	gonum "github.com/gonum/graph"
+	"github.com/openshift/origin/pkg/api/graph"
+	buildapi "github.com/openshift/origin/pkg/build/api"
+	buildutil "github.com/openshift/origin/pkg/build/util"
+	"github.com/openshift/origin/pkg/client"
+	"github.com/openshift/origin/pkg/cmd/dockerregistry"
+	deployapi "github.com/openshift/origin/pkg/deploy/api"
+	imageapi "github.com/openshift/origin/pkg/image/api"
+	"github.com/openshift/origin/pkg/image/registry/imagestreamimage"
+)
+
+// pruneAlgorithm contains the various settings to use when evaluating images
+// and layers for pruning.
+type pruneAlgorithm struct {
+	minimumAgeInMinutesToPrune int
+	tagRevisionsToKeep         int
+}
+
+// ImagePruneFunc is a function that is invoked for each image that is
+// prunable, along with the list of image streams that reference it.
+type ImagePruneFunc func(image *imageapi.Image, streams []*imageapi.ImageStream) []error
+
+// LayerPruneFunc is a function that is invoked for each registry, along with
+// a DeleteLayersRequest that contains the layers that can be pruned and the
+// image stream names that reference each layer.
+type LayerPruneFunc func(registryURL string, req dockerregistry.DeleteLayersRequest) (requestError error, layerErrors map[string][]error)
+
+// ImagePruner knows how to prune images and layers.
+type ImagePruner interface {
+	// Run prunes images and layers.
+	Run(imagePruneFunc ImagePruneFunc, layerPruneFunc LayerPruneFunc)
+}
+
+// imagePruner implements ImagePruner.
+type imagePruner struct {
+	g         graph.Graph
+	algorithm pruneAlgorithm
+}
+
+var _ ImagePruner = &imagePruner{}
+
+/*
+NewImagePruner creates a new ImagePruner.
+
```

Comment:
This is the right place to have a high-level comment describing the lack of a delete API in the registry and explaining why this is factored like it is.